common/src/main/java/com/wynntils/wynn/model/CharacterManager.java

```diff
@@ -80,14 +86,27 @@ public static void onWorldStateChanged(WorldStateEvent e) {
             inCharacterSelection = true;
         }
 
-        if (e.getNewState() == WorldStateManager.State.WORLD
-                && e.getOldState() != WorldStateManager.State.CHARACTER_SELECTION) {
-            // We went directly to a world without coming from the character selection
-            // menu. This means the player has "autojoin" enabled, and that we did not
-            // get a chance to read the character info from the character selection menu.
-            // Instead, we send a container query to read it from the character (compass) menu.
-            WynntilsMod.info("Scheduling character info query");
-            scanCharacterInfoPage();
+        if (e.getNewState() == WorldStateManager.State.WORLD) {
+            if (e.getOldState() != WorldStateManager.State.CHARACTER_SELECTION) {
+                // We went directly to a world without coming from the character selection
+                // menu. This means the player has "autojoin" enabled, and that we did not
+                // get a chance to read the character info from the character selection menu.
+                // Instead, we send a container query to read it from the character (compass) menu.
+                WynntilsMod.info("Scheduling character info query");
+
+                // This time, we need to scan character info and profession info as well.
+                scanCharacterInfoPage();
+            } else {
+                // We did not auto-join, we have a correct ID already.
+                int oldId = currentCharacter.id;
+                scanCharacterInfoPage();
+                currentCharacter = new CharacterInfo(
```

Comment:
Actually, is this correct? scanCharacterInfoPage() is only scheduling an async scan; you can not trust that any values has been updated when it returns. And in it we call:
`                    currentCharacter =
                            CharacterInfo.parseCharacterFromCharacterMenu(characterInfoItem, professionInfoItem, -1);`
so we will set the currentCharacter there anyway, and most likely overwriting what you are setting here.

Reply:
Yeah, this is bad. How do we fix it though, in a sensible way?

Reply:
I think what you want to do is to make `scanCharacterInfoPage` take in an `int id` which is either -1 or oldId, and use it instead.

And the entire if statement boils down to wether id should be set to -1 (we don't know), or currentCharacter.id (as in we do know).

Reply:
And maybe in the future rewrite this even more so we can dynamically update the CharacterInfo. But that's another PR.

Reply:
I think that now you can remove the `currentCharacter = new CharacterInfo(...)` from here, and just rely on the assignment in scanCharacterInfoPage.

Reply:
@kristofbolyai Not sure if this comment got lost in the torrent of other comments, but if you remove this reassignment, I think you're all good to go.

Reply:
I saw this, but I already turned off my PC yesterday. Doing this today.